### PR TITLE
session/statistics: discard feedbacks from `delete` / `update`

### PR DESCRIPTION
**Pull request by eurekaka**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/17438

Problem Summary:

Feedbacks from `update`, `delete` may be invalid.

### What is changed and how it works?

What's Changed:

Discard those feedbacks.

How it Works:

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

Side effects

N/A

### Release note 
- Discard feedbacks generated from `delete` / `update` statements.

### CODE REVIEW

File: session/session.go [hunk 1]
```diff
@@ -337,6 +337,10 @@ func (s *session) GetSessionManager() util.SessionManager {
 }
 
 func (s *session) StoreQueryFeedback(feedback interface{}) {
+	sc := s.sessionVars.StmtCtx
+	if sc.InUpdateStmt || sc.InDeleteStmt {
```
Comment (zz-jason) [inline session/session.go:341]: Should we also ignore feedback from `insert` and `replace` statements?

**Comment by zz-jason** [comment]
@eurekaka please also cherry-pick this change to release 2.1, 3.0, 3.1 and 4.0. Since query feedback is enabled by default on these branches.

**Comment by XuHuaiyu** [comment]
Should we also handle `insert select`?

File: executor/distsql.go [hunk 2]
```diff
@@ -245,7 +245,9 @@ func (e *IndexReaderExecutor) Close() error {
 		copStats := e.ctx.GetSessionVars().StmtCtx.RuntimeStatsColl.GetRootStats(e.plans[0].ExplainID().String())
 		copStats.SetRowNum(e.feedback.Actual())
 	}
-	e.ctx.StoreQueryFeedback(e.feedback)
+	if e.ctx.GetSessionVars().StmtCtx.InSelectStmt {
```
Comment (XuHuaiyu) [inline executor/distsql.go:248]: Why do we need to double-check here?
IMHO, we only need to decide whether to fetch feedback during planning?

File: statistics/feedback.go [hunk 3]
```diff
@@ -98,10 +98,17 @@ func (q *QueryFeedback) CalcErrorRate() float64 {
...
-func (q *QueryFeedback) CollectFeedback(numOfRanges int) bool {
+// 1: the feedback is not generated by select query;
+// 2: the histogram is nil or has no buckets;
+// 3: the number of scan ranges exceeds the limit because it may affect the performance;
+// 4: it does not pass the probabilistic sampler.
+func CollectFeedback(sc *stmtctx.StatementContext, q *QueryFeedback, numOfRanges int) bool {
+	if sc.RuntimeStatsColl != nil {
+		return true
+	}
+	if !sc.InSelectStmt {
```
[the remark this answers is not in the document]
Comment (eurekaka) [reply statistics/feedback.go:109]: > `sc.RuntimeStatsColl != nil` will be true when `explain analyze update`?

Yes, that's why we need to check `sc.RuntimeStatsColl` firstly.

File: executor/distsql.go (hunk 2 above)
Comment (eurekaka) [reply executor/distsql.go:248]: Same reason, `explain analyze` may generate feedbacks we don't want to dump.

File: statistics/feedback.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (zz-jason) [reply statistics/feedback.go:109]: I'm a little confused, can we simplify these two `if` statements to check whether the current statement is a `select` statement? Seems the first `if sc.RuntimeStatsColl != nil` can be removed?

[the remark this answers is not in the document]
Comment (eurekaka) [reply statistics/feedback.go:109]: `explain analyze` uses stats feedback for execution information collecting, so even if it is not a select query, e.g, `explain analyze update`, we should collect feedbacks for it? though `explain analyze update` would not output the operator infos now, maybe the behavior would change some day?

[the remark this answers is not in the document]
Comment (zz-jason) [reply statistics/feedback.go:109]: It depends on the behavior of `explain analyze` for DML statements. For example, if `explain analyze update` updates the table, I think we should not collect the query feedback in this scenario.

[the remark this answers is not in the document]
Comment (zz-jason) [reply statistics/feedback.go:109]: Usually, `explain analyze` used for query optimization in the staging or testing environment, it's hardly used in the application I think. Maybe it's acceptable to ignore the query feedback for these `explain analyze` statements.

**Pull request merged**